packages/@aws-cdk/aws-servicecatalog/lib/portfolio.ts

```diff
@@ -0,0 +1,351 @@
+    new CfnPortfolioPrincipalAssociation(this, `PortolioPrincipalAssociation${hashId}`, {
+      portfolioId: this.portfolioId,
+      principalArn: principalArn,
+      principalType: PrincipalType.IAM,
+    });
+  }
+
+  /**
+   * Add a launch notification constraint.
+   */
+  public addEventNotifications(props: EventNotificationsProps) {
+    AssociationManager.addLaunchNotificationConstraint(this, props);
+  }
+
+  /**
+   * Add a launch role constraint.
+   */
+  public addLaunchRole(props: LaunchRoleProps) {
+    AssociationManager.addLaunchRoleConstraint(this, props);
+  }
+
+  /**
+   * Add a launch template constraint.
+   */
+  public addProvisioningRules(props: ProvisioningRulesProps) {
+    AssociationManager.addTemplateConstraint(this, props);
+  }
+
+  /**
+   * Add a resource template constraint.
+   */
+  public allowTagUpdates(props: TagUpdatesProps) {
+    AssociationManager.addResourceUpdateConstraint(this, props);
+  }
+
+  /**
+   * Add a stack set template constraint.
+   */
+  public addStackSetConstraint(props: StackSetConstraintProps) {
+    AssociationManager.addStackSetConstraint(this, props);
+  }
+
+  /**
+   * Add tagOptions
+   * @param tagOption
+   */
+  public addTagOptions(tagOption: TagOption) {
+    AssociationManager.associateTagOption(this, this.portfolioId, tagOption);
+  }
+
+  /**
+   * Create a unique id based off the L1 portfolio.
+   */
+  protected abstract getKeyForPortfolio(value: string): string;
+}
+
+/**
+ * Properties for a Portfolio.
+ */
+export interface PortfolioProps {
+
+  /**
+   * Enforces a particular physical portfolio name.
+   */
+  readonly portfolioName: string;
+
+  /**
+   * The provider name.
+   */
+  readonly providerName: string;
+
+  /**
+   * The accept language.
+   * @default - No accept language provided
+   */
+  readonly acceptLanguage?: AcceptLanguage;
+
+  /**
+   * Description for portfolio.
+   * @default - No description provided
+   */
+  readonly description?: string;
+
+  /**
+   * A collection of tags attached to portfolio.
+   * @default - No tags provided
+   */
+  readonly tags?: Tag[];
+}
+
+/**
+ * A Service Catalog portfolio.
+ */
+export class Portfolio extends PortfolioBase {
+
+  /**
+   * Creates a Portfolio construct that represents an external portfolio.
+   *
+   * @param scope The parent creating construct (usually `this`).
+   * @param id The construct's name.
+   * @param attrs Portfolio import properties
+   */
+  public static fromPortfolioAttributes(scope: Construct, id: string, attrs: PortfolioAttributes): IPortfolio {
+    const parts = Stack.of(scope).parseArn(attrs.portfolioArn);
+    function throwError(errorMessage: string): never {
+      throw new Error(errorMessage);
+    }
+    class Import extends PortfolioBase {
+      public readonly id = id
+      public readonly portfolioId = parts.resourceName ?? throwError('Portfolio arn missing Portfolio ID during import from attributes');
```

Comment:
Please also include the value that caused the error in the message:

```suggestion
      public readonly portfolioId = parts.resourceName ?? throwError('Portfolio ARN is missing the Portfolio ID: ' + attrs.portfolioArn);
```